autopush/base.py

```diff
@@ -45,3 +45,33 @@ def write_error(self, code, **kwargs):
             self.log.failure("Error in handler: %s" % code,
                              client_info=self._client_info)
         self.finish()
+
+    def authenticate_peer_cert(self):
+        """Authenticate the client per the configured client_certs.
+
+        Aborts the request w/ a 401 on failure.
+
+        """
+        cert = self.request.connection.transport.getPeerCertificate()
+        assert cert, "Expected a TLS peer cert (VERIFY_FAIL_IF_NO_PEER_CERT)"
+
+        cert_signature = cert.digest('sha256')
+        cn = cert.get_subject().CN
+        auth = self.ap_settings.client_certs.get(cert_signature)
+        if auth is not None:
+            # TLS authenticated
+            self._client_info['tls_auth'] = auth
+            self._client_info['tls_auth_sha256'] = cert_signature
+            self._client_info['tls_auth_cn'] = cn
+            return
+
+        self.log.warn("Failed TLS auth",
+                      tls_failed_sha256=cert_signature,
+                      tls_failed_cn=cn,
+                      client_info=self._client_info)
+        self.set_status(401)
+        # "Transport mode" isn't standard, inspired by:
+        # http://www6.ietf.org/mail-archive/web/tls/current/msg05589.html
+        self.set_header('WWW-Authenticate',
+                        'Transport mode="tls-client-certificate"')
+        self.finish()
```

Comment:
Yep, finish flushes everything (headers, write buffers) and sets a flag causing any subsequent RequestHandler writes to raise a RuntimeError

When called from prepare() it stops the handler's target method from even running